### PR TITLE
refactor: new react app

### PR DESCRIPTION
**Pull request by frolic**
depends on https://github.com/latticexyz/mud/pull/261 and a new mud release with https://github.com/latticexyz/mud/pull/261 + https://github.com/latticexyz/mud/pull/253

- refactors app entry point as a pure react app and mount phaser/noa as hooks/components
- leans on vite's react plugin to do better HMR

### CODE REVIEW

File: packages/client/src/useNetworkLayer.ts
Comment (frolic) [inline packages/client/src/useNetworkLayer.ts]: rethink this, might be a mem leak for unresolved/skipped promises

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [inline packages/client/src/usePhaserLayer.ts]: This is how we should do the network layer disposing

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [inline packages/client/src/usePhaserLayer.ts]: may need a promise ref to ignore any slow-to-resolve promises that resolve after the current promise

File: packages/client/src/useNoaLayer.ts
Comment (frolic) [inline packages/client/src/useNoaLayer.ts]: refactor to be more like phaser layer

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [inline packages/client/src/usePhaserLayer.ts]: trigger reloading chunks?

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [inline packages/client/src/usePhaserLayer.ts]: trigger reloading chunks?

File: packages/client/src/layers/noa/systems/createInputSystem.ts
Comment (frolic) [inline packages/client/src/layers/noa/systems/createInputSystem.ts:145]: we'll eventually move this out

File: packages/client/src/layers/noa/systems/createInputSystem.ts
Comment (frolic) [inline packages/client/src/layers/noa/systems/createInputSystem.ts]: round these?

File: packages/client/src/layers/phaser/createPhaserLayer.ts
Comment (frolic) [inline packages/client/src/layers/phaser/createPhaserLayer.ts]: add logging library

File: packages/client/src/MapLayerToggle.tsx
Comment (alvrs) [inline packages/client/src/MapLayerToggle.tsx]: Should we move these components into a `components` folder? Actually - the react layer is probably not needed anymore, is it? We could just move all those components (only the ones we need) into a global component folder. We also don't need the stuff in layers/react/engine anymore I guess

File: packages/client/src/useObservable.ts
Comment (alvrs) [inline packages/client/src/useObservable.ts]: this seems to work re: default value
```suggestion
function useObservable<T>(observable: Observable<T>): T | undefined;
function useObservable<T>(observable: Observable<T>, defaultValue: T): T;

function useObservable<T>(observable: Observable<T>, defaultValue?: T) {
  const [value, setValue] = useState<T | undefined>(defaultValue);
  useEffect(() => {
    if (!observable) return;
    const subscription = observable.subscribe(setValue);
    return () => subscription.unsubscribe();
  }, [observable]);
  return value;
}
```

File: packages/client/src/useObservable.ts
Comment (alvrs) [inline packages/client/src/useObservable.ts]: should we call this `useObservableValue`? you're probably more familiar with hook naming conventions than I am

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [reply packages/client/src/usePhaserLayer.ts]: fixed in https://github.com/latticexyz/mud/pull/261/commits/50b56e9cc1ef64cd988f88ba2bd3cc70c2eaa1de

File: packages/client/src/useObservable.ts
Comment (alvrs) [reply packages/client/src/useObservable.ts]: Also, should we move this to MUD? Seems very useful for others. We could start a new `@latticexyz/mud-react` package (name tbd) as base for more useful generic react hooks/helpers, or just move it to `std-client/hooks.ts` for now

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [reply packages/client/src/usePhaserLayer.ts]: fixed in https://github.com/latticexyz/mud/pull/261/commits/50b56e9cc1ef64cd988f88ba2bd3cc70c2eaa1de

**Comment by frolic** [comment]
this is what logs look like after adding `debug` lib:
image

File: packages/client/vite.config.ts
Comment (frolic) [inline packages/client/vite.config.ts:50]: @alvrs do you have a sense for which of these config options we still use/need? esp when it comes time to replace parcel

File: packages/client/src/MapLayerToggle.tsx
Comment (alvrs) [inline packages/client/src/MapLayerToggle.tsx]:
```suggestion
```

File: packages/client/src/MapLayerToggle.tsx
Comment (alvrs) [inline packages/client/src/MapLayerToggle.tsx]:
```suggestion
```

File: packages/client/src/layers/noa/setup/setupNoaEngine.ts
Comment (alvrs) [inline packages/client/src/layers/noa/setup/setupNoaEngine.ts]:
```suggestion
```

File: packages/client/src/usePhaserLayer.ts
Comment (alvrs) [inline packages/client/src/usePhaserLayer.ts]: why do we set it to `RESIZE` in the config and then override it here?

File: packages/client/src/usePromiseValue.ts
Comment (alvrs) [inline packages/client/src/usePromiseValue.ts:1]: this also seems useful and general enough to be put in MUD (maybe the new react package, or maybe std-client or utils)

File: packages/client/vite.config.ts
Comment (alvrs) [reply packages/client/vite.config.ts:50]: at least `build` and `preview`. I think `resolve` and `optimizeDeps` were used to solve some bunding issue we ran into at some point, but we should probably approach approach the bunding situation with fresh eyes soon (given we can't even use vite for building for prod right now when mud is not linked locally). I believe `server` was used to fix an issue where vite complained it doesn't have access to the locally linked MUD (although it seems to work for me right now)

**Review by alvrs — COMMENTED** [review]
Looking great! Just some smaller nits

One general comment: in my experience `TODO:` comments are quickly forgotten after merging if not tracked elsewhere. Before merging we should either open an issue for them if we're actually planning on doing them, or removing the TODO comments.

File: packages/client/vite.config.ts
Comment (frolic) [reply packages/client/vite.config.ts:50]: ~all~ _mostly_ fixed! https://github.com/latticexyz/opcraft-readonly/pull/3/commits/9cd6fbb159dfa7bbffb439cff8e49928ae75539e

File: packages/client/src/MapLayerToggle.tsx
Comment (frolic) [reply packages/client/src/MapLayerToggle.tsx]: I think it makes sense to move some of this into a `react` directory to distinguish it from the other parts of the stack (phaser, noa, etc.) and drop some of the UI things into a `react/ui` directory or something. I was planning to take an organizational pass at the very end once things became more clear, can do in this PR or as a follow up.

I should note that I'm not a huge fan of the `components` directory pattern that you see in a lot of example apps/repos. I tend to structure things so that functional parts of the app are together, e.g. a set of components and hooks and utils/modules related to the map view are together in a `map` directory. And I tend to wait on organizing until it feels necessary rather than doing it too soon and having only one file in a directory.

File: packages/client/src/useObservable.ts
Comment (frolic) [reply packages/client/src/useObservable.ts]: Ohh interesting! I was trying to do this all inline, rather than an overloaded function def.

File: packages/client/src/useObservable.ts
Comment (frolic) [reply packages/client/src/useObservable.ts]: happy to move it but been blocked on mud PR merges/version bumps so I may leave it here for now

I'll plan to PR some of the react stuff into mud and migrate to it once its landed/released

**Comment by frolic** [comment]
> One general comment: in my experience `TODO:` comments are quickly forgotten after merging if not tracked elsewhere. Before merging we should either open an issue for them if we're actually planning on doing them, or removing the TODO comments.

I agree with this, and will open issues for the TODOs that are important and out of scope of this PR. But I also like to leave TODOs in code as a reminder when it comes time to revisit a thing, refactor, etc. It's much easier to see TODOs and comments about trade-offs/unknowns in the code that's about to change than to go find any issues related to a piece of code.

File: packages/client/src/useObservable.ts
Comment (frolic) [reply packages/client/src/useObservable.ts]: > should we call this `useObservableValue`? you're probably more familiar with hook naming conventions than I am

I actually started with `useLastObservableValue` and then `useObservableValue` and then `useObservable`, because I couldn't of any other use case for putting an observable into a hook and getting something out of it. Not opposed to `useObservableValue` if you think it's clearer!

File: packages/client/src/usePhaserLayer.ts
Comment (frolic) [reply packages/client/src/usePhaserLayer.ts]: `RESIZE` seemed like a reasonable default, but since we're gonna replace the config value anyway, I'll get rid of that one.

I think leaving the `NONE` scale mode here makes the most sense, to tightly couple it to the resize handling behavior in the hook.

File: packages/client/src/MapLayerToggle.tsx
Comment (frolic) [reply packages/client/src/MapLayerToggle.tsx]: here's what I ended up with:
image

~everything related to showing UI over the app layers (phaser/noa) is in `ui`

File: packages/client/src/useObservable.ts
Comment (frolic) [reply packages/client/src/useObservable.ts]: ~Looks like this is more like a type override and TS doesn't give me warnings/hints when writing the function body itself.~  Nevermind! Looks like it does tell me when the function/implementation signature is not compatible with the defined overloads. 👍 

Would still be nice to be able to do this inline but can't figure out how...

File: packages/client/vite.config.ts
Comment (frolic) [inline packages/client/vite.config.ts:21]: this is causing OOM in CI, may need to disable there or figure out how to bump memory needs

File: packages/client/src/useObservable.ts
Comment (frolic) [reply packages/client/src/useObservable.ts]: ended up going with `useObservableValue` to match `usePromiseValue`

File: packages/client/src/app/PhaserLayer.tsx
Comment (alvrs) [inline packages/client/src/app/PhaserLayer.tsx:11]: Great habit to explain design decisions in a comment for future reference!

File: packages/client/src/layers/noa/systems/createInputSystem.ts
Comment (alvrs) [reply packages/client/src/layers/noa/systems/createInputSystem.ts:145]: #4 (for reference)

File: packages/client/src/layers/phaser/systems/createInputSystem.ts
Comment (alvrs) [inline packages/client/src/layers/phaser/systems/createInputSystem.ts:67]: planned in this PR or in a followup?

**Review by alvrs — APPROVED** [review]
Looks great! Very excited for these patterns and utils to eventually make its way into mud & default react templates!

File: packages/client/src/layers/phaser/systems/createInputSystem.ts
Comment (frolic) [reply packages/client/src/layers/phaser/systems/createInputSystem.ts:67]: yep!

**Pull request merged**
